### ISSUE TITLE
Compiler Issue: Early Return Fails When Using Unsafe Types 

### ISSUE THREAD

**Issue by spotandjake** [grain-lang/grain#1697]
Early return seems to fail when using an unsafe type other then `wasmI32`,
Code that causes the issue:
```ts
module Main

include "runtime/dataStructures"
include "runtime/numbers"
include "runtime/unsafe/wasmf32"
include "runtime/unsafe/wasmi32"
from DataStructures use { newFloat32 }
@unsafe
let rf = z => { // Rational approximation of (asin(x)-x)/x^3
  from WasmF32 use { mul as (*), add as (+), sub as (-), div as (/) }
  // see: musl/src/math/asinf.c and SUN COPYRIGHT NOTICE above
  let pS0 = WasmF32.reinterpretI32(0x3E2AAA75n) //  1.6666586697e-01f
  let pS1 = WasmF32.reinterpretI32(0xBD2F13BAn) // -4.2743422091e-02f
  let pS2 = WasmF32.reinterpretI32(0xBC0DD36Bn) // -8.6563630030e-03f
  let qS1 = WasmF32.reinterpretI32(0xBF34E5AEn) // -7.0662963390e-01f

  let p = z * (pS0 + z * (pS1 + z * pS2))
  let q = 1.0w + z * qS1
  return p / q
}
@unsafe
let internal_asin = x => {
  // see: musl/src/math/asinf.c and SUN COPYRIGHT NOTICE above
  // Operators
  from WasmI32 use { and as (&), eq as (==), geU as (>=), ltU as (<) }
  from WasmF32 use { mul as (*), add as (+), sub as (-), div as (/) }
  // Constants
  let const_pio2 = WasmF32.reinterpretI32(
    0x3FC90FDBn
  ) // 1.570796326794896558e+00f
  let const_Ox1p_120f = WasmF32.reinterpretI32(0x03800000n) // 0x1p-120f
  // Calcs
  let sx = x
  let hx = WasmI32.reinterpretF32(x) & 0x7FFFFFFFn
  if (hx >= 0x3F800000n) {
    if (hx == 0x3F800000n) {
      return x * const_pio2 + const_Ox1p_120f
    } else {
      return 0.0w / (x - x)
    }
  } else if (hx < 0x3F000000n) {
    if (hx < 0x39800000n && hx >= 0x00800000n) {
      return x
    } else {
      return x + x * rf(x * x)
    }
  }
  let z = 0.5w - WasmF32.abs(x) * 0.5w
  let s = WasmF32.sqrt(z) // sic
  let x = const_pio2 - 2.0w * (s + s * rf(z))
  return WasmF32.copySign(x, sx)
}
@unsafe
let asin = x => {
  let x = Numbers.coerceNumberToWasmF32(x)
  return WasmI32.toGrain(newFloat32(internal_asin(x))): Number
}
// Output
print(asin(0.5))
```

![image](https://user-images.githubusercontent.com/40705786/220762034-a3db3708-f2db-4924-8a1f-017eea8b1436.png)
![image](https://user-images.githubusercontent.com/40705786/220762080-d622af5e-c818-470e-92b3-9711ec004e6e.png)

I could be very wrong in diagnoses here but i think it may have something todo with this line:
https://github.com/grain-lang/grain/blob/49a399dbcfc24957a3875ae094ad0a2c4b636a32/compiler/src/middle_end/anf_helper.re#L179

I think that may need to take an allocation type in that is dependent on the type used by the user.

**Comment by spotandjake** [comment]
Here is a much smaller example with a much more helpful error segment 
![image](https://user-images.githubusercontent.com/40705786/220808352-23f55317-e7c4-462a-b2ea-244e8fb8ea1e.png)
```ts
module Main

@unsafe
let x = () => {
  return 1.0W
}

@unsafe
let y = x()
```
compiled with: `grain test.gr --no-gc --no-pervasives`

![image](https://user-images.githubusercontent.com/40705786/220808668-39944179-ae4e-4667-a114-e561b33e0047.png)

If i understand this right the issue isnt the function return type but the block built within the functions return type. isnt correct.

**Comment by spotandjake** [comment]
I did a tiny bit more rigging
![image](https://user-images.githubusercontent.com/40705786/220825226-15ae53e3-d1e2-4b00-aa78-4459a6124ac8.png)
and that is the compiled block for the early return on a WasmF64 and the block on an i32 looks like.
![image](https://user-images.githubusercontent.com/40705786/220825340-2ed15be5-00ce-4c16-a592-c9b98456cfd3.png)
and on a number it looks like:
![image](https://user-images.githubusercontent.com/40705786/220825370-68d19fdd-c67a-45d8-ae9d-d816871680bf.png)
so you can see each time the `return` is dropped. and then we return a `void` from the block which gets returned the issue is the `void` type and `return` type's type signatures are fighting the void doesnt really need to exist though given it is dead code.

**Comment by spotandjake** [comment]
This seems to have been fixed, Likely in #1714 as a side effect of `return` being refactored
